crates/obj/src/patchable.rs

```diff
@@ -0,0 +1,342 @@
+    // TODO support all platforms, but now don't fix unsupported.
+    let mut file = match parse_elf_object_mut(bytes) {
+        Ok(file) => file,
+        Err(_) => {
+            return Ok(());
+        }
+    };
+
+    ensure!(
+        file.e_phoff() != 0 && file.e_phnum() == 1,
+        "program header table must created"
+    );
+
+    let shstrtab_off = shstrtab_off(file.as_mut());
+
+    if let Some(mut section) = ElfSectionIterator::new(file.as_mut())
+        .find(|s| is_text_section(bytes, shstrtab_off, s.as_ref()))
+    {
+        // Patch vaddr, and save file location and its size.
+        section.sh_addr_set(code_region.0 as u64);
+    }
+
+    // LLDB wants segment with virtual address set, placing them at the end of ELF.
+    let ph_off = file.e_phoff();
+    let (v_offset, size) = code_region;
+    let mut program = file.program_at(ph_off);
+    program.p_vaddr_set(v_offset as u64);
+    program.p_paddr_set(v_offset as u64);
+    program.p_memsz_set(size as u64);
+
+    Ok(())
+}
+
+/// Removes all patched information from loadable file fields.
+/// Inverse of `patch_loadable_file`.
+pub fn sanitize_loadable_file(bytes: &mut [u8]) -> Result<(), Error> {
+    const NON_EXISTENT_RANGE: (*const u8, usize) = (std::ptr::null(), 0);
+    patch_loadable_file(bytes, NON_EXISTENT_RANGE)
+}
+
+fn is_text_section(bytes: &mut [u8], shstrtab_off: u64, section: &dyn ElfSection) -> bool {
+    if section.sh_type() != SHT_PROGBITS {
+        return false;
+    }
+    // It is a SHT_PROGBITS, but we need to check sh_name to ensure it is our function
+    let sh_name_off = section.sh_name();
+    let sh_name = unsafe {
+        CStr::from_ptr(
+            bytes
+                .as_ptr()
+                .offset((shstrtab_off + sh_name_off as u64) as isize) as *const c_char,
+        )
+        .to_str()
+        .expect("name")
+    };
+    sh_name == ".text"
+}
+
+fn shstrtab_off(file: &mut dyn ElfObject) -> u64 {
+    ElfSectionIterator::new(file)
+        .rev()
+        .find(|s| s.sh_type() == SHT_STRTAB)
+        .map_or(0, |s| s.sh_offset())
+}
+
+fn parse_elf_object_mut<'data>(obj: &'data [u8]) -> Result<Box<dyn ElfObject>, Error> {
+    let ident: &Ident = unsafe { &*(obj.as_ptr() as *const Ident) };
+    let elf = match (ident.class, ident.data) {
+        (ELFCLASS64, ELFDATA2LSB) => Box::new(ElfObject64LE {
+            header: unsafe { &mut *(obj.as_ptr() as *mut FileHeader64<_>) },
+        }),
+        (c, d) => {
+            bail!("Unsupported elf format (class: {}, data: {})", c, d);
+        }
+    };
+    match elf.e_machine() {
+        EM_X86_64 | EM_ARM => (),
+        machine => {
+            bail!("Unsupported ELF target machine: {:x}", machine);
+        }
+    }
+    let e_shentsize = elf.e_shentsize();
+    ensure!(e_shentsize as usize == elf.target_shentsize(), "size of sh");
+    Ok(elf)
+}
+
+trait ElfObject {
+    fn e_ident(&self) -> &Ident;
```

Comment:
It looks like some of these reader methods are [already on a predefined trait](https://docs.rs/object/0.20.0/object/read/elf/trait.FileHeader.html#tymethod.e_ident), could that be used instead?

(it looks like setters aren't found there though)

Reply:
That is the main reason for code duplication -- I can add a comment about setters.